### DISCUSSION TITLE
Understanding connection lifetime & pooling

### DISCUSSION THREAD

**Discussion by pkese** [DarkWanderer/ClickHouse.Client#458] [Q&A]
A appologize in advance, this is not an issue but more of a question.

I read the page about [connection lifetime & pooling](https://github.com/DarkWanderer/ClickHouse.Client/wiki/Connection-lifetime-&-pooling) and it is rather technical, e.g. to understand what is going on, one would ideally have deep understanding what `System.Net.Http.HttpClient` does, how does it do pooling internally and how it gets used by ClickHouse.Client.

What I'd like to know is simply: if I have an open connection, what happens if I pass the same connection to two async tasks, that might use that same connection concurrently?

Will in such case ...
a) all hell break lose,
b) or it will run but operations will be passed to the database sequentially,
c) or it will run fine with requests passed to the database concurrently?

Ideally, a short paragraph about that would be nice also in the above document (implementation information is useful for hardcore dotnet developers, but not many people are familiar with all the nitty gritty technical details and would benefit more with a clear explanation of what happens and what the best use pattern is).

Thanks. I love this driver.

**Comment by DarkWanderer** [comment]
Hi. Sure, no problem.

`CickHouseConnection` uses `HttpClient` internally and hence generally follows same rules - for example, you can queue multiple requests to `HttpClient` from different threads, and the will be executed in parallel, but - with default settings - no more than 2 simultaneous requests to same server at the same time. All the 'extra' queries will wait in queue until there is room to execute. See [MaxConnectionsPerServer](https://learn.microsoft.com/en-us/dotnet/api/system.net.http.httpclienthandler.maxconnectionsperserver?view=net-8.0) documentation for further info

2 notable exceptions are:

- if sessions are enabled, a specially constructed `HttpClient` instance is used internally which only allows one parallel query at the time. This is to adhere to ClickHouse mechanics of `session_id`. However, in that case there would not be a global limit - there would be active HTTP connections up to number of `ClickHouseConnection` instances
- If custom `HttpClient` or `IHttpClientFactory` is passed, then the behavior will depend on the settings of passed object

Other than that, there is no transient state in the connection object - as long as settings are not changed, it can be considered thread-safe

So speaking of best practice patterns, I would say:

- Use default settings (just pass ConnectionString) for generic UI use. Number of connection objects is not limited and there is reasonable 'pooling' with 2 connections by default
- If you need to use sessions, make sure you use a limited number of Connection objects, as there is potentially large number of connections which can be opened
- For advanced use - if you need to tweak parameters heavily - create a custom implementation of [IHttpClientFactory](https://learn.microsoft.com/en-us/dotnet/api/system.net.http.ihttpclientfactory) - as Connection object will call `httpClientFactory.CreateClient` to create short-lived `HttpClient` instances to handle requests
- Passing a 'static' `HttpClient` instance is also available, just make sure to set `Timeout` property to desired value as it can't be changed later
